[FIX] [rocBLAS][Tensile] enabled for all GFX103X (RDNA2) archs

umarinkovic opened the pull request:
## Motivation

I created a patch for the purposes of TheRock and the maintainers there asked me to land the changes here instead.

## Technical Details

Officially, rocBLAS is supposed to support the gfx1030 and probably gfx1031. I've added the rest of these cards to the CMake and Python verification logic, and I've also updated the enumerations used in library-loading / kernel-dispatch which caused most gfx103X-libraries to produce failures in rocBLAS smoke-tests (gfx1030 being the only one that actually could pass the tests)

## Test Plan

I've built rocBLAS for these architectures using TheRock on both Windows and Linux, and I've ran the rocBLAS smoke-tests on both the gfx1031 and gfx1036. I've also tested the libraries generated for other cards on these two.

## Test Result

[----------] Global test environment tear-down
[==========] 4514 tests from 169 test suites ran. (28963 ms total)
[  PASSED  ] 4514 tests.
rocBLAS version: 5.1.0.590d2b4f5e
rocBLAS-commit-hash: 590d2b4f5e898701ec9be7504dd8ff5854658b93
Tensile-commit-hash: 

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

TorreZuk commented:
@umarinkovic  the smoke test is only a quick API consistency test not a good test at all, please run psdb and osdb test sets.   rocblas_rtest.py -t psdb and then with -t osdb.   If you prefer gtest filters that is just "*quick*:*pre_checkin*" and then "*nightly*".

umarinkovic commented:
> @umarinkovic the smoke test is only a quick API consistency test not a good test at all, please run psdb and osdb test sets. rocblas_rtest.py -t psdb and then with -t osdb. If you prefer gtest filters that is just "_quick_:_pre_checkin_" and then "_nightly_".

Hmm, yeah, makes sense. I was wondering why there were only smoke-tests available, I guess that that's enough for the general purposes of TheRock.

I'll make sure to run more extensive tests and report back with the results.

umarinkovic commented:
Small update: ran the 
>psdb and osdb test sets. rocblas_rtest.py -t psdb and then with -t osdb

`psdb` tests hang at the beginning, and osdb tests pass successfully on Windows but have however failed on Linux with 16 failures in blas1 dotc_batched / strided_batched tests.

This is going to need some more work, I'll start by updating the yaml configs for the remaining architectures.

TorreZuk commented on projects/rocblas/CMakeLists.txt:
```diff
@@ -80,12 +80,12 @@ list( APPEND CMAKE_PREFIX_PATH ${ROCM_PATH}/llvm ${ROCM_PATH} ${ROCM_PATH}/hip /
+  set( TARGET_LIST_ROCM_6.0 "gfx900;gfx906:xnack-;gfx908:xnack-;gfx90a:xnack+;gfx90a:xnack-;gfx942;gfx1010;gfx1012;gfx1030;gfx1031;gfx1032;gfx1033;gfx1034;gfx1035;gfx1036;gfx1100;gfx1101;gfx1102")
+  set( TARGET_LIST_ROCM_6.3 "gfx900;gfx906:xnack-;gfx908:xnack-;gfx90a;gfx942;gfx1010;gfx1012;gfx1030;gfx1031;gfx1032;gfx1033;gfx1034;gfx1035;gfx1036;gfx1100;gfx1101;gfx1102;gfx1151;gfx1200;gfx1201")
+  set( TARGET_LIST_ROCM_7.0 "gfx900;gfx906:xnack-;gfx908:xnack-;gfx90a;gfx942;gfx950;gfx1010;gfx1012;gfx1030;gfx1031;gfx1032;gfx1033;gfx1034;gfx1035;gfx1036;gfx1100;gfx1101;gfx1102;gfx1150;gfx1151;gfx1200;gfx1201")
+  set( TARGET_LIST_ROCM_7.1 "gfx900;gfx906:xnack-;gfx908:xnack-;gfx90a;gfx942;gfx950;gfx1010;gfx1012;gfx1030;gfx1031;gfx1032;gfx1033;gfx1034;gfx1035;gfx1036;gfx1100;gfx1101;gfx1102;gfx1103;gfx1150;gfx1151;gfx1200;gfx1201")
```
I think we are leaning toward wanting these removed from default lists, otherwise I think the changes are good if tested.  Defaults can be discussed in separate PR

umarinkovic replied:
I've removed the added architectures from the default lists. Running tests locally, I've ran into some problems:

I've tested on both Windows and Linux on the gfx1031, with the following results:

```
python rocblas_rtest.py -t osdb

[----------] Global test environment tear-down
[==========] 611934 tests from 171 test suites ran. (7437760 ms total)
[  PASSED  ] 611934 tests.
```

```
python rocblas_rtest.py -t psdb:

[----------] 244 tests from _/asum
[----------] 244 tests from _/asum (1104 ms total)

[----------] 304 tests from _/asum_batched # hangs here
Python Exception: Traceback (most recent call last):
  File "C:\dev\repos\TheRock\out\build\dist\rocm\bin\rocblas_rtest.py", line 234, in run_cmd
    output = test_proc.stdout.readline()
  File "C:\Python313\Lib\encodings\cp1252.py", line 22, in decode
    def decode(self, input, final=False):

KeyboardInterrupt

FAILED xml test suite!
```

As you can see, the _-osdb_ aka _nightly_ tests all pass, while the _-psdb_ aka _quick:pre-checkin_ hang on the _asum_batched suite.

There's also a small error in the `rocblas_rtest.py` script for Windows that I needed to fix for which I can open a separate PR. 

On Linux, it's a different story. I tested both the gfx1031 and the gfx1100 as a control, and the results were practically the same:

gfx1031:
```
python rocblas_rtest.py -t osdb

16 FAILED TESTS
[ SKIPPED  ] 9 tests.
[ PASSED   ] 611973 tests.
[ FAILED   ] 16 tests.
rocBLAS version: 5.1.0.f005caf7c3
rocBLAS-commit-hash: a3cfc707edd382d33afaa8fcf6fc5053049869cc
Tensile-commit-hash:
hipBLASLt version: 1.1.0 commit-hash: f005caf7c3
command line: rocblas-test --gtest_output=xml --gtest_color=yes --gtest_filter=*nightly*-*known_bug*
FAILED xml test suite!
```
Some of the errors that show up here:
```
[0;32m[----------] [m80 tests from _/dot_batched
SIGNAL raised in: blas1/nightly_dot_batched_kernel_test_f32_c_256_n3_n3_10002_0
/therock/src/rocm-libraries/projects/rocblas/clients/common/gtest_helpers.cpp:175: Failure
Failed
Received Segmentation fault signal
################################################################################
[0;32m[----------] [m40 tests from _/gemv_strided_batched
Warning: hip can't allocate 12303495872 bytes (11 GB)
hip OutOfMemory at /therock/src/rocm-libraries/projects/rocblas/clients/common/../include/blas2/testing_gemv_strided_batched.hpp:402
hipGetLastError at end of test: blas2/nightly_gemv_strided_batched_very_large_f64_c_N_16010_16010_2_16010_256320100_2_32020_2_2_32020_3
rocBLAS error from hip error code: 'hipErrorOutOfMemory':2
Warning: hip can't allocate 15008075136 bytes (13 GB)
hip OutOfMemory at /therock/src/rocm-libraries/projects/rocblas/clients/common/../include/blas2/testing_gemv_strided_batched.hpp:402
hipGetLastError at end of test: blas2/nightly_gemv_strided_batched_very_large_f64_r_N_25020_25020_2_25020_625000400_2_50040_2_2_50040_3
rocBLAS error from hip error code: 'hipErrorOutOfMemory':2
Warning: hip can't allocate 15008075136 bytes (13 GB)
hip OutOfMemory at /therock/src/rocm-libraries/projects/rocblas/clients/common/../include/blas2/testing_gemv_strided_batched.hpp:402
hipGetLastError at end of test: blas2/nightly_gemv_strided_batched_very_large_f32_c_N_25020_25020_2_25020_625000400_2_50040_2_2_50040_3
rocBLAS error from hip error code: 'hipErrorOutOfMemory':2
Warning: hip can't allocate 30016150272 bytes (27 GB)
hip OutOfMemory at /therock/src/rocm-libraries/projects/rocblas/clients/common/../include/blas2/testing_gemv_strided_batched.hpp:402
hipGetLastError at end of test: blas2/nightly_gemv_strided_batched_very_large_f64_c_N_25020_25020_2_25020_625000400_2_50040_2_2_50040_3
rocBLAS error from hip error code: 'hipErrorOutOfMemory':2
```

```
python rocblas_rtest.py -t psdb

8 FAILED TESTS
[ SKIPPED  ] 2 tests.
[ PASSED   ] 590556 tests.
[ FAILED   ] 8 tests.
rocBLAS version: 5.1.0.f005caf7c3
rocBLAS-commit-hash: a3cfc707edd382d33afaa8fcf6fc5053049869cc
Tensile-commit-hash:
hipBLASLt version: 1.1.0 commit-hash: f005caf7c3
command line: rocblas-test --gtest_output=xml --gtest_color=yes --gtest_filter=*quick*:*pre_checkin*-*known_bug*
FAILED xml test suite!
```

Again, same types of errors as in the previous case: Segmentation fault and hipOutOfMem.

I ran these tests on my Linux machine using the Radeon RX 7900 XTX (gfx1100) as a control group, and I've experienced these same issues. 

Are these tests reliable? I snooped around the workflows in this repo and I only found the .yaml blueprints for component tests, not the actual runs. (I'm very new to github actions so it's possible I just missed something). Is it possible to run the tests on a remote machine in case my setup is faulty? This is how I ran the tests after building the entire ROCm stack through TheRock:

```
docker run -it --rm \
  --device=/dev/kfd \
  --device=/dev/dri \
  --group-add video \
  --cap-add=SYS_PTRACE \
  --security-opt seccomp=unconfined \
  -v ./:/therock/src \
  -v ./output-linux-portable:/therock/output \
  -w /therock/src \
  ghcr.io/rocm/therock_build_manylinux_x86_64:main \
  /bin/bash -c "
    # run setup 
    
    pip install -r requirements.txt && \
    pip install -r requirements-test.txt && \
    mkdir -p /opt/rocm/bin && \
    cp -r /therock/output/build/dist/rocm /opt && \
    export THEROCK_BIN_DIR=/opt/rocm/bin && \
    export PATH=$PATH:$THEROCK_BIN_DIR && \

    # Drop into shell
    exec bash"
```
after running ^: `cd /opt/rocm/bin && python rocblas_rtest.py -osdb`

TorreZuk replied:
Reduced these gemv tests in yaml and merged.  export ROCBLAS_CLIENT_RAM_GB_LIMIT=32 can avoid host out of memory which should allow for HMM reductions as well.  They should move to SKIPPED

TorreZuk replied:
Conversation continued so will resolve this one.

TorreZuk commented:
Okay the gemv out of memory are fine to ignore if low VRAM test node.  I am removing in https://github.com/ROCm/rocm-libraries/pull/1823 but blocked by unrelated CI failures in the rock.  If you install AOCL as reference before doing theRock style build we would look for all tests passing, you probably can't use their build if it doesn't install a supported reference library for all tests.   Tests that fail for an enabled gfx that isn't fully supported are a discussion point to raise with a new issue.  Otherwise if Tensile team approves (fix the static analysis failure)  we will likely approve and merge in near future.  Making such a list a default list (what you just removed) is another issue to raise.  Thanks for your patience.

umarinkovic commented:
>Okay the gemv out of memory are fine to ignore if low VRAM test node. I am removing in https://github.com/ROCm/rocm-libraries/pull/1823 but blocked by unrelated CI failures in the rock. If you install AOCL as reference before doing theRock style build we would look for all tests passing, you probably can't use their build if it doesn't install a supported reference library for all tests.

Here are the test results of building with AOCL on Ubuntu:

```shell
rocblas-test --gtest_output=xml --gtest_color=yes --gtest_filter=*nightly*-*known_bug*
rocBLAS info: client (OPENMP) reduced omp_set_num_threads to 110
rocBLAS warning: LD_LIBRARY_PATH override may use incompatible rocblas
rocBLAS info: Using reference library '/opt/AMD/aocl/aocl-linux-aocc-5.1.0/aocc/lib_ILP64/libblis-mt.a'
rocBLAS warning: Reference library may not support 64-bit input arguments. If running a test suite, please use --gtest_filter=-*stress* to avoid 64-bit test failures.
rocBLAS version: 5.1.0.
rocBLAS-commit-hash:
Tensile-commit-hash:
hipBLASLt version: 1.1.0 commit-hash:
Query device success: there are 1 devices
-------------------------------------------------------------------------------
Device ID 0 : AMD Radeon RX 6700 XT gfx1031
with 12.9 GB memory, max. SCLK 2855 MHz, max. MCLK 1000 MHz, memoryBusWidth 24 Bytes, compute capability 10.3
maxGridDimX 2147483647, sharedMemPerBlock 65.5 KB, maxThreadsPerBlock 1024, warpSize 32
-------------------------------------------------------------------------------
[0;32m[  PASSED  ] [m611994 tests.
[0;32m[  SKIPPED ] [m4 tests, listed below:
[0;32m[  SKIPPED ] [m_/gemv_batched.blas2/nightly_gemv_batched_very_large_f64_r_N_25020_25020_2_25020_2_2_2_3
[0;32m[  SKIPPED ] [m_/gemv_strided_batched.blas2/nightly_gemv_strided_batched_very_large_f64_r_N_25020_25020_2_25020_625000400_2_50040_2_2_50040_3
[0;32m[  SKIPPED ] [m_/gemv_strided_batched.blas2/nightly_gemv_strided_batched_very_large_f32_c_N_25020_25020_2_25020_625000400_2_50040_2_2_50040_3
[0;32m[  SKIPPED ] [m_/gemv_strided_batched.blas2/nightly_gemv_strided_batched_very_large_f64_c_N_25020_25020_2_25020_625000400_2_50040_2_2_50040_3
[ SKIPPED  ] 4 tests.
[ PASSED   ] 611994 tests.
[ FAILED   ] 0 tests.
```

Pre-checkin tests have more failures now...:

```shell
rocBLAS info: client (OPENMP) reduced omp_set_num_threads to 110
rocBLAS warning: LD_LIBRARY_PATH override may use incompatible rocblas 
rocBLAS info: Using reference library '/opt/AMD/aocl/aocl-linux-aocc-5.1.0/aocc/lib_ILP64/libblis-mt.a'
rocBLAS warning: Reference library may not support 64-bit input arguments. If running a test suite, please use --gtest_filter=-*stress* to avoid 64-bit test failures.
rocBLAS version: 5.1.0.
rocBLAS-commit-hash: 
Tensile-commit-hash: 
hipBLASLt version: 1.1.0 commit-hash: 
Query device success: there are 1 devices
-------------------------------------------------------------------------------
Device ID 0 : AMD Radeon RX 6700 XT gfx1031
with 12.9 GB memory, max. SCLK 2855 MHz, max. MCLK 1000 MHz, memoryBusWidth 24 Bytes, compute capability 10.3
maxGridDimX 2147483647, sharedMemPerBlock 65.5 KB, maxThreadsPerBlock 1024, warpSize 32
-------------------------------------------------------------------------------
42 FAILED TESTS
[ SKIPPED  ] 2 tests.
[ PASSED   ] 590522 tests.
[ FAILED   ] 42 tests.
```

Example of a failure:
```shell
[0;31m[  FAILED  ] [m_/gemv_batched.blas2/pre_checkin_gemv_batched_NaN_f64_r_C_32_32_nnan_32_2_nnan_2_9, where GetParam() = { function: "gemv_batched", name: "gemv_batched_NaN", category: "pre_checkin", known_bug_platforms: "", alpha: .nan, beta: .nan, stride_a: 1024, M: 32, N: 32, lda: 32, incx: 2, incy: 2, batch_count: 9, a_type: f64_r, b_type: f64_r, c_type: f64_r, d_type: f64_r, compute_type: f64_r, initialization: rand_int, gpu_arch: "", flush_batch_count: 1, transA: 'C' }
 (2 ms)
/therock/src/rocm-libraries/projects/rocblas/clients/common/../include/unit.hpp:489: Failure
Value of: rocblas_isnan(hGPU[k][i + idx])
  Actual: false
Expected: true

/therock/src/rocm-libraries/projects/rocblas/clients/common/../include/unit.hpp:489: Failure
Value of: rocblas_isnan(hGPU[k][i + idx])
  Actual: false
Expected: true
```

TorreZuk commented:
Okay if the failures are only gemv NaN tests that is an issue with AOCL 5.1 I have reported to them to fix, you can go back to 5.0 or 4.2 and also use the gcc build not the aocc if you want to be consistent with us.

TorreZuk commented:
@davidd-amd and @yoichiyoshida  review please

TinyServal commented:
Thank you for working on this, I have access to gfx1035 and can help with testing if needed.

umarinkovic commented:
> Thank you for working on this, I have access to gfx1035 and can help with testing if needed.

Hey, that sounds great! Are you on Linux or on Windows?

TinyServal commented:
Linux, Debian 13 specifically. Was trying to build TheRock which obviously didn't work, looked it up and the various issues and pull requests led me here.

TorreZuk commented:
Additional comment some tests above used "/opt/AMD/aocl/aocl-linux-aocc-5.1.0/aocc/lib_ILP64/libblis-mt.a" but should have used the gcc version as mentioned in documentation.   Last reported gemv NaN won't occur with AOCL gcc 4.2

umarinkovic commented:
@TorreZuk 

Hey, sorry to bother you, but could you ping the others for a review once more? There are a few issues on **TheRock** blocked by this as well. Thanks 🙌

fjankovi commented on shared/tensile/Tensile/Common.py:
```diff
@@ -75,60 +77,96 @@ class DeveloperWarning(Warning):
-globalParameters["PerformanceMetric"] = "DeviceEfficiency" # performance metric for benchmarking; one of {DeviceEfficiency, CUEfficiency}
-globalParameters["PrintLevel"] = 1                # how much info to print in generator. 0=none, 1=standard, 2=add code commands, 3=verbose
-globalParameters["ClientLogLevel"] = 3            # the log level of client. 0=Error, 1=Terse, 2=Verbose, 3=Debug (Aligned with ResultReporter.hpp)
+globalParameters[
```
This PR should not be reformatting this file.

TorreZuk reviewed (approved):
rocBLAS side is fine, @yoichiyoshida who on Tensile side can review this?

umarinkovic commented:
> This PR should not be reformatting shared/tensile/Tensile/Common.py, shared/tensile/Tensile/AsmCaps.py or any other file in general.

Thanks for the review, I reverted the formatting commit.

umarinkovic commented:
> @umarinkovic Can you add changes for gfx1033 and gfx1036 in the tensile/pytest.ini file? I also think that yaml files for mentioned targets are missing in projects/rocblas/library/src/blas3/Tensile/Logic/asm_full folder. You need to copy navi21 folder, name the folder according to the mapping in architectureMap (Common.py file) and create changes in the yaml files (ISA, target name)

Hey, thanks for the review. I've added the missing yaml files and modified the pytest.ini file in line with your comment.

TorreZuk commented:
Enabled pipelines and re-running windows build as the error looks like it wasn't getting merged to tip of develop.  May require rebase if same error occurs.  Tensile static analysis (formatting) must pass as well.

umarinkovic commented:
> Enabled pipelines and re-running windows build as the error looks like it wasn't getting merged to tip of develop. May require rebase if same error occurs. Tensile static analysis (formatting) must pass as well.

>This PR should not be reformatting shared/tensile/Tensile/Common.py, shared/tensile/Tensile/AsmCaps.py or any other file in general.

I'm not sure I understand the formatting requirements.

TorreZuk commented:
```
sh -c '$CLANG_FORMAT_DIR/clang-format -style=file ./Tensile/Source/lib/include/Tensile/PlaceholderLibrary.hpp | diff - ./Tensile/Source/lib/include/Tensile/PlaceholderLibrary.hpp'
110c110
<             return "TensileLibrary_*_gfx1036";
---
>             return "TensileLibrary_*_gfx1036";   
```
Must be whitespace on that line

umarinkovic commented:
> ...
> Must be whitespace on that line

Yep, I ran clang-format against the other Tensile code files just in case, it was just those three white spaces. Sorry, I'm not too familiar with formatting.

umarinkovic commented:
@TorreZuk

> Enabled pipelines and re-running windows build as the error looks like it wasn't getting merged to tip of develop. May require rebase if same error occurs.

There was a conflict when applying the old Tensile-formatting commit which was reverted in the next commit anyway. I've rebased on top of develop excluding those two commits, should be good now.

umarinkovic commented:
@TorreZuk 

Hi :wave:, any updates on this PR?

TorreZuk commented:
Hi @umarinkovic I only have permission to run CI on this as there were failures.  Will speak to a gardener today who has authority to merge with failures and see if we can merge if failures match baseline.  I enabled CI and re-running windows build as last failure was in compile but looks like it was false failure.

TorreZuk commented:
@amcamd failures match baseline so think we can merge.
